### PR TITLE
Revamp config compare for compatibility with more configs

### LINKED ISSUE #622: Comparing two sh configs is broken.
The `compare_config` method on the `Config` base class assumes a section/key/value structure and is used in the CLI/API-facing `config.tools.compare_configs()` function. 

Using two very basic identical sh files containing:
```
foo=bar
bar=banana
baz=2
```

We see the following stack trace:

```
$ uw config compare --file-1-path test.sh --file-2-path test2.sh 
[2024-10-16T12:22:11]     INFO - test.sh
[2024-10-16T12:22:11]     INFO + test2.sh
[2024-10-16T12:22:11]     INFO ---------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/christina.holt/conda/envs/DEV-uwtools/bin/uw", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/christina.holt/work/uwtools/src/uwtools/cli.py", line 93, in main
    sys.exit(0 if modes[args[STR.mode]](args) else 1)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/christina.holt/work/uwtools/src/uwtools/cli.py", line 199, in _dispatch_config
    return actions[args[STR.action]](args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/christina.holt/work/uwtools/src/uwtools/cli.py", line 208, in _dispatch_config_compare
    return uwtools.api.config.compare(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/christina.holt/work/uwtools/src/uwtools/api/config.py", line 35, in compare
    return _compare(
           ^^^^^^^^^
  File "/Users/christina.holt/work/uwtools/src/uwtools/config/tools.py", line 38, in compare_configs
    return cfg_1.compare_config(cfg_2.data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/christina.holt/work/uwtools/src/uwtools/config/formats/base.py", line 175, in compare_config
    for key, val in items.items():
                    ^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'items'
```

We should amend the logic in `Config.compare_config` to handle any depth of config.

### PR DESCRIPTION
**Pull request by maddenp-noaa**
**Synopsis**

Fixes #622.

In addition to fixing the compare-two-identical-`.sh`-configs example from the ticket, `.sh` configs with differences are now handled, with differences shown using a familiar diff syntax:

`a.sh`
```
foo=bar
bar=banana
baz=2
```
`b.sh`
```
foo=bar
bar=papaya
```
```
$ uw config compare --file-1-path a.sh --file-2-path b.sh
[2024-11-14T23:47:32]     INFO - a.sh
[2024-11-14T23:47:32]     INFO + b.sh
[2024-11-14T23:47:32]     INFO ---------------------------------------------------------------------
[2024-11-14T23:47:32]     INFO ↓ ? = info | -/+ = line unique to - or + file | blank = matching line
[2024-11-14T23:47:32]     INFO ---------------------------------------------------------------------
[2024-11-14T23:47:32]     INFO - bar: banana
[2024-11-14T23:47:32]     INFO ?      ^ ^ ^
[2024-11-14T23:47:32]     INFO + bar: papaya
[2024-11-14T23:47:32]     INFO ?      ^ ^ ^
[2024-11-14T23:47:32]     INFO - baz: '2'
[2024-11-14T23:47:32]     INFO   foo: bar
```

Also, depth > 2 configs, which also were not previously processed, now work:

`a.yaml`
```
foo:
  bar:
    baz: qux
```
`b.yaml`
```
foo:
  bar:
    baz: asdf
```
```
$ uw config compare --file-1-path a.yaml --file-2-path b.yaml
[2024-11-14T23:50:03]     INFO - a.yaml
[2024-11-14T23:50:03]     INFO + b.yaml
[2024-11-14T23:50:03]     INFO ---------------------------------------------------------------------
[2024-11-14T23:50:03]     INFO ↓ ? = info | -/+ = line unique to - or + file | blank = matching line
[2024-11-14T23:50:03]     INFO ---------------------------------------------------------------------
[2024-11-14T23:50:03]     INFO   foo:
[2024-11-14T23:50:03]     INFO     bar:
[2024-11-14T23:50:03]     INFO -     baz: qux
[2024-11-14T23:50:03]     INFO +     baz: asdf
```

Mixed-depth configs work:
`c.yaml`
```
foo:
  bar: qwerty
```
```
$ uw config compare --file-1-path a.yaml --file-2-path c.yaml
[2024-11-14T23:51:55]     INFO - a.yaml
[2024-11-14T23:51:55]     INFO + c.yaml
[2024-11-14T23:51:55]     INFO ---------------------------------------------------------------------
[2024-11-14T23:51:55]     INFO ↓ ? = info | -/+ = line unique to - or + file | blank = matching line
[2024-11-14T23:51:55]     INFO ---------------------------------------------------------------------
[2024-11-14T23:51:55]     INFO   foo:
[2024-11-14T23:51:55]     INFO +   bar: qwerty
[2024-11-14T23:51:55]     INFO -   bar:
[2024-11-14T23:51:55]     INFO -     baz: qux
```

Namelist configs are compared via their YAML representations:
`a.nml`
```
&a s="foo" n=999 /
```
`b.nml`
```
&a
  s="foo"
  n=1
/
```
```
$ uw config compare --file-1-path a.nml --file-2-path b.nml
[2024-11-14T23:56:13]     INFO - a.nml
[2024-11-14T23:56:13]     INFO + b.nml
[2024-11-14T23:56:13]     INFO ---------------------------------------------------------------------
[2024-11-14T23:56:13]     INFO ↓ ? = info | -/+ = line unique to - or + file | blank = matching line
[2024-11-14T23:56:13]     INFO ---------------------------------------------------------------------
[2024-11-14T23:56:13]     INFO   a:
[2024-11-14T23:56:13]     INFO -   n: 999
[2024-11-14T23:56:13]     INFO ?      ^^^
[2024-11-14T23:56:13]     INFO +   n: 1
[2024-11-14T23:56:13]     INFO ?      ^
[2024-11-14T23:56:13]     INFO     s: foo
```

Unit tests are updated/expanded, and docs are updated.

This may be considered a breaking change because the logged output is different.

**Type**

- [x] Bug fix (corrects a known issue)
- [x] Documentation
- [x] Enhancement (adds new functionality)

**Impact**

- [x] This is a breaking change (changes existing functionality)

**Checklist**

- [x] I have added myself and any co-authors to the PR's _Assignees_ list.
- [x] I have reviewed the documentation and have made any updates necessitated by this change.

### CODE REVIEW

File: docs/sections/user_guide/cli/drivers/upp/help.out [hunk 1]
```diff
@@ -12,6 +12,8 @@ Optional arguments:
 
 Positional arguments:
   TASK
+    control_file
+      The GRIB control file
```
Comment (maddenp-noaa) [inline docs/sections/user_guide/cli/drivers/upp/help.out:16]: This must have been missed in a previous update -- it is unrelated to the changes in this PR.

File: src/uwtools/config/formats/base.py [hunk 2]
```diff
@@ -76,6 +79,26 @@ def _characterize_values(self, values: dict, parent: str) -> tuple[list, list]:
...
+        return sio.getvalue().splitlines(keepends=True)
+
+    @staticmethod
+    def _compare_config_log_header() -> None:
+        """
+        Log a visual header and description of diff markers.
+        """
+        log.info("-" * MSGWIDTH)
+        log.info("? => info, -/+ => line only in - or + file, blank => matching line")
+        log.info("-" * MSGWIDTH)
```
Comment (maddenp-noaa) [inline src/uwtools/config/formats/base.py:100]: These two methods are just helpers for `compare_config()` -- code factored out of that method to keep it reasonably abstract.

File: src/uwtools/config/formats/base.py [hunk 3]
```diff
@@ -158,43 +181,35 @@ def _parse_include(self, ref_dict: Optional[dict] = None) -> None:
 
     # Public methods
 
-    def compare_config(self, dict1: dict, dict2: Optional[dict] = None) -> bool:
+    @abstractmethod
+    def as_dict(self) -> dict:
+        """
+        Returns a pure dict version of the config.
+        """
```
Comment (maddenp-noaa) [inline src/uwtools/config/formats/base.py:188]: This is mainly to support `NMLConfig`, whose `.data` attribute can be an `f90nml` `Namelist` object which, although `dict`-like, is not actually a `dict` and runs into problems in `yaml.dump()`. `NMLConfig` now implements `as_dict()` in such a way (see below) that it returns a pure-`dict` object. The other config types just return their `.data` attribute, which is already a `dict`.

File: src/uwtools/tests/config/formats/test_fieldtable.py [hunk 4]
```diff
@@ -49,3 +69,25 @@ def test_fieldtable_simple(config, ref, tmp_path):
...
+    FieldTableConfig(d).dump(path)
+    with open(path, "r", encoding="utf-8") as f:
+        assert f.read().strip() == expected
+
+
+def test_fieldtable_dump_dict(dumpkit):
+    d, expected, path = dumpkit
+    FieldTableConfig.dump_dict(d, path=path)
+    with open(path, "r", encoding="utf-8") as f:
+        assert f.read().strip() == expected
```
Comment (maddenp-noaa) [inline src/uwtools/tests/config/formats/test_fieldtable.py:93]: Here and below, I added unit tests for the new `as_dict()` method, as well as for the existing `dump()` and `dump_dict()` methods. The latter two were already transitively tested, but now have their own direct tests, too.

File: src/uwtools/tests/config/formats/test_base.py [hunk 5]
```diff
@@ -129,15 +133,65 @@ def test_compare_config(caplog, fmt, salad_base):
...
+    expected = """
+    ---------------------------------------------------------------------
+    ? => info, -/+ => line only in - or + file, blank => matching line
+    ---------------------------------------------------------------------
+      salad:
+        base: kale
+    -   dressing: balsamic
+    ?             ^  ^ ^^^
+    +   dressing: italian
+    ?             ^^  ^ ^
```
Comment (christinaholtNOAA) [inline src/uwtools/tests/config/formats/test_base.py:147]: I appreciate that it thinks the `a`s and `l`s are the same.

File: src/uwtools/config/formats/base.py [hunk 6]
```diff
@@ -168,33 +199,16 @@ def compare_config(self, dict1: dict, dict2: Optional[dict] = None) -> bool:
...
+        dict2 = self.as_dict() if dict2 is None else dict2
+        lines1, lines2 = map(self._compare_config_get_lines, [dict1, dict2])
+        difflines = list(difflib.ndiff(lines2, lines1))
+        if all(line[0] == " " for line in difflines):  # i.e. no +/-/? lines
+            return True
+        if header:
+            self._compare_config_log_header()
+        for diffline in difflines:
+            log.info(diffline.rstrip())
+        return False
```
Comment (christinaholtNOAA) [inline src/uwtools/config/formats/base.py:211]: Wow.

**Review by christinaholtNOAA — APPROVED** [review]
Geez! You really went above and beyond on the bug fix. The user experience upgrade and the clean solution are fantastic!

File: src/uwtools/config/formats/base.py (hunk 6 above)
Comment (maddenp-noaa) [reply src/uwtools/config/formats/base.py:211]: Concise, right? It helps to reduce a new problem to an existing one with a known solution (useful diff algorithms).

**Pull request merged**